fix: infer "release" automatically if "config.release" not specified

rchl opened the pull request:
Use Sentry CLI to automatically deduce "release" configuration if not
specified manually.

It's necessary to store the release in a generated file at build-time,
and import that file at run-time, so that the server-side Sentry can use
it on initialization (which happens within the module itself rather than
within a server-side plugin).

Resolves #200

rchl commented:
I can use the hook to generate the options but I also need to know when **not generating** (so that I only read the options at run time) and that I can only do by checking that property.

pimlie commented:
What about doing the inverse, so always load the options (if they exist) but unset it once you hit `build:before` or something. Wouldnt that have the same effect?

rchl commented:
Always loading means they would be loaded also at build time and we don't want that as that would load options generated on previous build.

I thought of various solutions but I couldn't see hook based solutions that would work correctly.

rchl commented:
If you have concrete suggestion with concrete code or pseudo code then I'll look into that possibility.

pimlie commented:
> Always loading means they would be loaded also at build time and we don't want that as that would load options generated on previous build.

```
moduleFn() {
 - load runtime options
 - add ready (or listen) hook 
   - do server init in ready hook
 - add build:before hook
   - set/load build/dev options
     - during this overwrite/unset runtime options
   - addPlugins/addTemplates
   - do server init if ready hook wasnt called yet (not totally sure if they run in sync or not and too lazy atm to look it up)
}

rchl commented:
I've changed it to use the hooks and moved all relevant code into them. It won't be very easy to review probably.

For reference, here is the order the hooks are run:
```
yarn build
HOOK run:before
HOOK config
HOOK modules:before
Sentry reporting is enabled (client side: enabled, server side: enabled)
HOOK modules:done
HOOK ready
HOOK build:before
HOOK builder:prepared
HOOK builder:extendPlugins
HOOK build:extendRoutes
HOOK build:templates
HOOK webpack:config
Enabling uploading of release sourcemaps to Sentry
HOOK build:compile
HOOK build:compiled
HOOK build:resources
HOOK build:resources
HOOK build:compile
HOOK bundler:done
HOOK build:compiled
HOOK build:resources
HOOK build:resources
HOOK build:done

yarn start
HOOK run:before
HOOK config
HOOK modules:before
Sentry reporting is enabled (client side: enabled, server side: enabled)
HOOK modules:done
HOOK render:before
HOOK render:resourcesLoaded
HOOK render:setupMiddleware
HOOK render:errorMiddleware
HOOK render:done
HOOK ready
HOOK listen
```

Doing server init in `ready` hook wasn't possible as that hook runs very early when building and we need to figure out "release" from the `build:before` first.

pimlie commented on lib/module.js:
```diff
@@ -243,6 +249,28 @@ export default async function SentryModule (moduleOptions) {
     })
   }
+
+  async function initSentry () {
```
Including functions within other functions should be avoided as much as possible cause this makes them impossible to unit test. Just use function arguments and trust that the GC is smart enough to release the ref once the fn isnt needed anymroe

rchl commented on lib/module.js:
```diff
@@ -223,12 +88,153 @@ export default async function SentryModule (moduleOptions) {
+    logger.info(`Sentry reporting is disabled (${why})`)
+  }
+
+  const moduleContainer = this
```
Not necessary. I have added this for older code where it was needed.

rchl replied:
Change to exported functions that gets passed `this`.

rchl commented:
> To improve readability & testability (is that a word?) I think it would be nicer to make all hook functions just separate functions that are also exported by the module (or a lib). Then the module file itself would only contain some imports, the default Config and calling the hooks which would provide much better insights into the lifecycle of this module

I've made those a standalone functions that gets passed arguments.
Those are still in the same file. I can move them somewhere else, just need to figure out good file name for those :)

rchl commented:
All good now @pimlie ?
